src/DotVVM.Framework/Compilation/CompiledAssemblyCache.cs

```diff
@@ -24,12 +102,44 @@ private Assembly DefaultOnResolving(System.Runtime.Loader.AssemblyLoadContext as
         /// <summary>
         /// Tries to resolve compiled assembly.
         /// </summary>
-        public Assembly TryResolveAssembly(object sender, ResolveEventArgs args)
+        public Assembly? TryResolveAssembly(object sender, ResolveEventArgs args)
         {
-            Assembly assembly;
-            return cachedAssemblies.TryGetValue(args.Name, out assembly) ? assembly : null;
+            return cachedAssemblies.TryGetValue(args.Name, out var assembly) ? assembly : null;
+        }
+#endif
+
+        public Assembly[] GetReferencedAssemblies()
+        {
+            return cachedAssemblies.Values.ToArray();
         }
+
+        public Assembly[] GetAllAssemblies()
+        {
+            if (configuration.ExperimentalFeatures.ExplicitAssemblyLoading.Enabled)
+            {
+                return GetReferencedAssemblies();
+            }
+            else
+            {
+#if DotNetCore
+                // auto-loads all referenced assemblies recursively
+                return DependencyContext.Default.GetDefaultAssemblyNames().Select(Assembly.Load).ToArray();
+#else
+                // this doesn't load new assemblies, but it is done in InvokeStaticConstructorsOnAllControls
+                return AppDomain.CurrentDomain.GetAssemblies();
 #endif
+            }
+        }
+
+        public bool IsAssemblyNamespace(string fullName)
+            => GetAllNamespaces().Contains(fullName, StringComparer.Ordinal);
```

Comment:
Hmm, I know this is just copy paste from the old utils. However, it would make more sense to create the `HashSet` once with the `Ordinal` comparer, and then cache it in a `Lazy` field.

Reply:
Fixed.